packages/wonder-blocks-layout/components/media-layout.js

```diff
@@ -52,104 +67,67 @@ type State = {|
 // fall back to this state.
 const DEFAULT_SIZE = "large";
 
+type CombinedProps = {|...Props, ...Context|};
+
 /**
  * `MediaLayout` is responsible for changing the rendering of contents at
  * differently sized viewports.  `MediaLayoutContext.Provider` can be used
  * to specify different breakpoint configurations.  By default it uses
  * `MEDIA_DEFAULT_SPEC`.  See media-layout-context.js for additiional options.
  */
-export default class MediaLayout extends React.Component<Props, State> {
-    watchHandlers: ?{
-        [query: string]: any,
-    };
-
-    static WATCHERS: {
-        [query: string]: any,
-    } = {};
-
-    state = {
-        size: undefined,
-    };
+class MediaLayoutInternal extends React.Component<CombinedProps, State> {
+    cleanupThunks: Array<() => void>;
+
+    constructor(props: CombinedProps) {
+        super(props);
+        this.state = {
+            size: undefined,
+        };
+        this.cleanupThunks = [];
+    }
 
-    componentWillUnmount() {
-        // We go through the component and remove all of the listeners
-        // that getCurrentSize attached.
-        for (const query of Object.keys(MediaLayout.WATCHERS)) {
-            const watcher = MediaLayout.WATCHERS[query];
-
-            if (watcher && this.watchHandlers) {
-                const {watchHandlers} = this;
-                const handler = this.watchHandlers[query];
-                watcher.removeListener(handler);
-                delete watchHandlers[query];
+    componentDidMount() {
+        // TODO(WB-534): handle changes to mediaSpec prop
+        for (const [size, spec] of ((Object.entries(
+            this.props.mediaSpec,
+        ): any): Array<[MediaSize, {query: string}]>)) {
+            const mql = MEDIA_QUERY_LISTS[spec.query];
+            // during SSR there are no MediaQueryLists
+            if (!mql) {
+                continue;
             }
+            const listener = (e) => {
+                if (e.matches) {
+                    this.setState({size});
+                }
+            };
+            mql.addListener(listener);
+            this.cleanupThunks.push(() => mql.removeListener(listener));
         }
     }
 
-    getCurrentSize(spec: MediaSpec) {
+    componentWillUnmount() {
+        this.cleanupThunks.map((thunk) => thunk());
```

Comment:
This is great, though I think semantically, you should use `forEach` rather than `map` as the latter implies the results matter and creates an array of the results which we're then discarding.

Also, what do you think about adding a comment that says something like "Remove our listeners"? I wonder if folks reading the code will know what the thunks are.